client/src/main/java/org/eclipse/hono/client/impl/HonoConnectionImpl.java

```diff
@@ -279,8 +292,47 @@ private void checkConnected(final Handler<AsyncResult<Void>> resultHandler) {
         if (isConnectedInternal()) {
             resultHandler.handle(Future.succeededFuture());
         } else {
-            resultHandler.handle(Future.failedFuture(
-                    new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+            if (connecting.get()) {
+                // connect attempt in progress - let its completion complete the resultHandler here
+                log.debug("connection attempt to server [{}:{}] in progress, connection check will be completed with its result",
+                        connectionFactory.getHost(), connectionFactory.getPort());
+                final Promise<Void> promiseToAdd = Promise.promise();
+                promiseToAdd.future().setHandler(resultHandler);
+                // atomically add the promise to checkConnectedPromises - but only if checkConnectedPromises hasn't been cleared already
+                final List<Promise<Void>> newCheckConnectedPromises = checkConnectedPromises
+                        .accumulateAndGet(Collections.singletonList(promiseToAdd), (existing, toAdd) -> {
+                            // no modification of the existing list done here, keeping the accumulatorFunction function side-effect free as required
+                            if (existing == null) {
+                                return null;
+                            }
+                            final List<Promise<Void>> promises = new ArrayList<>(existing.size() + 1);
+                            promises.addAll(existing);
+                            promises.add(toAdd.get(0));
+                            return promises;
+                        });
+                if (newCheckConnectedPromises == null) {
+                    // checkConnectedPromises wasn't updated with our promiseToAdd because the checkConnectedPromises were cleared in between
+                    resultHandler.handle(Future.failedFuture(
+                            new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+                }
+            } else {
+                resultHandler.handle(Future.failedFuture(
+                        new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "not connected")));
+            }
+        }
+    }
+
+    private void removeAndSucceedCheckConnectedPromises() {
+        final List<Promise<Void>> promises = checkConnectedPromises.getAndSet(null);
+        if (promises != null) {
+            promises.forEach(Promise::complete);
```

Comment:
I am a little concerned about invoking all of these handlers here from the same (current) handler on the event loop (Don't block the event loop!). This is most likely not a problem for just a few handlers but since we do not know what the handlers of the promises actually do, it might be better to schedule each completion separately, i.e.:

```
promises.forEach(promise -> context.runOnContext((promise.complete()));
```
WDYT?

Reply:
Yes, it's probably better. To be taken into account here is that `context` will usually be `null` in case of a failed connection attempt.